src/include/rho/GCNodeAllocator.hpp

```diff
@@ -0,0 +1,223 @@
+
+      /**
+       * \brief Allocate an object of at least the given size.
+       *
+       * This should only be used to allocate memory for rho::GCNode objects.
+       */
+      static void* allocate(std::size_t bytes);
+
+      /** \brief Apply function to all current allocations. */
+      static void applyToAllAllocations(std::function<void(void*)> f);
+
+      /** \brief Free a previously allocated object. */
+      static void free(void* p);
+
+      /** \brief Must be called before any allocations can be made. */
+      static void initialize();
+
+      /**
+       * \brief Find heap allocation start pointer.
+       *
+       * This function finds the corresponding allocation for an internal or
+       * exact object pointer.
+       *
+       * If the pointer is inside the bounds of the small object arena, then
+       * the corresponding superblock is found via pointer manipulation.
+       *
+       * If the pointer is inside the heap bounds, then we iterate over all hash
+       * collisions for the pointer to find the corresponding allocation.
+       */
+      static GCNode* lookupPointer(void* candidate);
+
+      /** \brief Print allocator state summary for debugging. */
+      static void printSummary();
+
+    private:
+      friend class AllocatorSuperblock;
+      friend class AllocationTable;
+      friend struct FreeListNode;
+
+      /** Allocation table for medium and large allocations. */
+      static AllocationTable* s_alloctable;
+
+      /** Number of different object sizes plus one for the small-object superblocks. */
+      static constexpr unsigned s_num_small_pools = (256 / 8) + 1;
+
+      /** Number of different object sizes plus one for the medium-object superblocks. */
+      static constexpr unsigned s_num_medium_pools = 18;
+
+      /**
+       * The freelists must fit all small superblock size classes, plus
+       * all large allocation sizes (= 33 + 64).
+       */
+      static constexpr unsigned s_num_freelists = s_num_small_pools + 64;
+
+      /** Free list heads, indexed by size class. */
+      static FreeListNode* s_freelists[s_num_freelists];
+
+      /** Pointers to superblocks with untouched blocks, indexed by size class. */
+      static AllocatorSuperblock* s_superblocks[s_num_small_pools + s_num_medium_pools];
+
+      /** Smallest known heap address. */
+      static uintptr_t s_heap_start;
+
+      /** Largest known heap address. */
+      static uintptr_t s_heap_end;
+
+      /**
+       * Create a (in place) freelist node for an alloacation and insert in a
+       * freelist by size class.
+       * This should not be used to insert superblock allocations in a freelist
+       * because it nulls out the superblock pointer.
+       */
+      static void addToFreelist(uintptr_t data, unsigned size_class);
+
+      /**
+       * \brief Add a free list node to a free list by size class.
+       */
+      static void addToFreelist(FreeListNode* free_node, unsigned size_class);
```

Comment:
Having two functions with the same name, but different semantics (at least for superblocks) is error-prone.